### PR TITLE
Soccer API (MVP) #1

### PR DESCRIPTION
**Pull request by nghiapham1026**
1. Scraped match data from espn.com to /getDataByDates:
Return the names of teams
Return the logos of each team
Return the date of games
Return the score of games
Return the match status of finished games: FT for finished games, elapsed time for live games, and time for upcoming games
2. Get all the matches 1 day from today until the user given end date, the default end date is 7 days from today.
Create function in dates.js to define and format dates
Format date strings: "YYYYMMDD"
Screenshot Capture - 2023-09-03 - 21-48-18

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,27 +1,30 @@
...
-const mongoose = require("mongoose");
-const todoRoutes = require("./routes/todo.js");
+const scrapeController = require("./controllers/webscrapeController");
+const { getDateRange } = require("./models/dates");
 
-main()
-  .then((res) => console.log(res))
-  .catch((err) => console.log(err));
+const { startDate, endDate } = getDateRange();
+const mongoose = require("mongoose");
```
Comment (saravpereira) [inline index.js]: all imports need to go together first then other function calls

File: index.js [hunk 2]
```diff
@@ -1,27 +1,30 @@
...
 
 async function main() {
-  await mongoose.connect(process.env.DB);
-}
+  try {
```
Comment (saravpereira) [inline index.js]: why is all the api wrapped in try catch?
This becomes a problem if we don't depend on DB for api

File: controllers/constants.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+    headlineUrl: "https://www.espn.com",
+    scoresURL: 'https://www.espn.com/soccer/scoreboard',
+  };
```
Comment (saravpereira) [inline controllers/constants.js]: const espnUrl = "https://www.espn.com"
const espnScoreUrl = `${espnUrl}/soccer/scoreboard`

module.exports = { espnUrl, espnScoreUrl };

File: index.js [hunk 4]
```diff
@@ -1,27 +1,30 @@
...
 
-app.use(bodyParser.json());
-app.use(bodyParser.urlencoded({ extended: false }));
+    app.use("/api/todo", require("./routes/todo.js"));
 
-app.get("/", (req, res) => {
-  res.send("hello world");
-});
+    app.get("/getDataByDates", (req, res) => {
+      scrapeController.scrapeEspn(startDate, endDate, req, res);
```
Comment (saravpereira) [inline index.js]: users should be able to pass an argument of end date. Right now const { startDate, endDate } = getDateRange(); 

Rethink this logic

Comment (saravpereira) [reply index.js]: consider passing startdate as a param that is YYYYMMDD that way you don't have to get date range unless it's not provided

File: index.js [hunk 5]
```diff
@@ -1,27 +1,30 @@
...
+    const app = express();
 
-app.use(bodyParser.json());
-app.use(bodyParser.urlencoded({ extended: false }));
+    app.use("/api/todo", require("./routes/todo.js"));
 
-app.get("/", (req, res) => {
-  res.send("hello world");
-});
+    app.get("/getDataByDates", (req, res) => {
```
Comment (saravpereira) [inline index.js]: consider renaming this api
/espn/getGamesByDates

**Comment by saravpereira** [comment]
I cannot run this to verify the api, what are the set up requirements for puppeteer?
Getting this error:

```
Error scraping: Error: Could not find Chrome (ver. 116.0.5845.96). This can occur if either
 1. you did not perform an installation before running the script (e.g. `npm install`) or
 2. your cache path is incorrectly configured (which is: /Users/sarapereira/.cache/puppeteer).
For (2), check out our guide on configuring puppeteer at https://pptr.dev/guides/configuration.
    at ChromeLauncher.resolveExecutablePath (/Users/sarapereira/FieldsAndGoals-server/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:300:27)
    at ChromeLauncher.executablePath (/Users/sarapereira/FieldsAndGoals-server/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:182:25)
    at ChromeLauncher.computeLaunchArguments (/Users/sarapereira/FieldsAndGoals-server/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:98:37)
    at async ChromeLauncher.launch (/Users/sarapereira/FieldsAndGoals-server/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:79:28)
    at async Object.scrapeEspn (/Users/sarapereira/FieldsAndGoals-server/controllers/webscrapeController.js:6:21)
```

**Comment by saravpereira** [comment]
this branch has some conflicting files fyi, you need to do a git pull and resolve the conflicts then push your changes again.
This is because you branched out of a test branch that never got merged and had the older version of the code.
Please **branch out of main and redo this work given the feedback** 
You may keep this pr open as reference before you open a new pr

**Review by saravpereira — CHANGES_REQUESTED** [review]
See comments in this pr

**Comment by nghiapham1026** [comment]
> I cannot run this to verify the api, what are the set up requirements for puppeteer? Getting this error:
> 
> ```
> Error scraping: Error: Could not find Chrome (ver. 116.0.5845.96). This can occur if either
>  1. you did not perform an installation before running the script (e.g. `npm install`) or
>  2. your cache path is incorrectly configured (which is: /Users/sarapereira/.cache/puppeteer).
> For (2), check out our guide on configuring puppeteer at https://pptr.dev/guides/configuration.
>     at ChromeLauncher.resolveExecutablePath (/Users/sarapereira/FieldsAndGoals-server/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:300:27)
>     at ChromeLauncher.executablePath (/Users/sarapereira/FieldsAndGoals-server/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:182:25)
>     at ChromeLauncher.computeLaunchArguments (/Users/sarapereira/FieldsAndGoals-server/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:98:37)
>     at async ChromeLauncher.launch (/Users/sarapereira/FieldsAndGoals-server/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:79:28)
>     at async Object.scrapeEspn (/Users/sarapereira/FieldsAndGoals-server/controllers/webscrapeController.js:6:21)
> ```

Run npm i axios puppeteer to install all dependencies

**Comment by nghiapham1026** [comment]
> this branch has some conflicting files fyi, you need to do a git pull and resolve the conflicts then push your changes again. This is because you branched out of a test branch that never got merged and had the older version of the code. Please **branch out of main and redo this work given the feedback** You may keep this pr open as reference before you open a new pr

Resolved all conflicts

File: index.js [hunk 6]
```diff
@@ -1,27 +1,28 @@
...
+  const app = express();
 
-app.use(bodyParser.json());
-app.use(bodyParser.urlencoded({ extended: false }));
+  app.use("/api/todo", require("./routes/todo.js"));
 
-app.use("/api/todo", todoRoutes);
+  app.get("/espn/getGamesByDates", (req, res) => {
+    scrapeController.scrapeEspn(startDate, endDate, req, res);
+  });
```
Comment (saravpereira) [inline index.js]: rethink this logic... how will users pass endDate if this api doesn't allow them to? 
getDateRange is hardcoding the endDate, which is not acceptable.
As I said in the previous comment, this should be driven by url parameters to the api call

File: index.js [hunk 7]
```diff
@@ -1,27 +1,28 @@
...
 
-app.get("/", scrapeController.scrapeEspn);
+  app.listen(8080, () => {
+    console.log("Server listening on port 8080");
+  });
+}
```
Comment (saravpereira) [inline index.js:34]: why is all the code wrapped inside the main function?

**Comment by saravpereira** [comment]
Still getting the same error as above when running api... **all dependencies are already installed**
You are missing this file, please add the changes so it won't break for others 
Screenshot 2023-09-05 at 12 09 48 PM
Pupeteer is used because Cheerio often returns null when scraping links and body texts

**Review by saravpereira — CHANGES_REQUESTED** [review]
see comments!

File: index.js [hunk 8]
```diff
@@ -1,27 +1,29 @@
...
-  .then((res) => console.log(res))
-  .catch((err) => console.log(err));
+mongoose.connect(process.env.DB);
+console.log("Connected to the database");
+
+const app = express();
 
-async function main() {
-  await mongoose.connect(process.env.DB);
-}
```
Comment (saravpereira) [inline index.js:16]: this change can be kept, why is this removed.
async function main() {
  await mongoose.connect(process.env.DB);
}

File: index.js [hunk 9]
```diff
@@ -1,27 +1,29 @@
 const express = require("express");
-const bodyParser = require("body-parser");
 require("dotenv").config();
 const scrapeController = require("./controllers/webscrapeController");
-
-const app = express();
+const { getDateRange } = require("./models/dates");
 const mongoose = require("mongoose");
```
Comment (saravpereira) [inline index.js]: all imports of external libraries go first then internal import

File: index.js [hunk 10]
```diff
@@ -1,27 +1,29 @@
...
+app.use("/api/todo", require("./routes/todo.js"));
 
-app.use(bodyParser.json());
-app.use(bodyParser.urlencoded({ extended: false }));
+app.get("/espn/getGamesByDates", (req, res) => {
+  /**
+   * User can specify startDate and endDate by appending them as query parameters:
+   * http://localhost:8080/espn/getGamesByDates?startDate=20230901&endDate=20230905
+   */
+  const userStartDate = req.query.startDate;
```
Comment (saravpereira) [inline index.js]: this api is only to get future games, so start date is always 1 day from today. Users should not be setting start date.
endDate should not be set for any time before 1 day from today -- if it is, throw an error

File: controllers/webscrapeController.js [hunk 11]
```diff
@@ -1,21 +1,107 @@
-const axios = require("axios");
-const cheerio = require("cheerio");
-const { url } = require("./constants");
+const puppeteer = require("puppeteer");
+const { espnUrl, espnScoreUrl } = require("./constants");
+const axios = require('axios');
```
Comment (saravpereira) [inline controllers/webscrapeController.js]: some imports are not used here, pl remove

File: controllers/webscrapeController.js [hunk 12]
```diff
@@ -1,21 +1,106 @@
-const axios = require("axios");
-const cheerio = require("cheerio");
-const { url } = require("./constants");
+const puppeteer = require("puppeteer");
+const { espnUrl, espnScoreUrl } = require("./constants");
```
Comment (saravpereira) [inline controllers/webscrapeController.js]: espnUrl is not used here, you can remove

File: controllers/webscrapeController.js [hunk 13]
```diff
@@ -1,21 +1,106 @@
...
 
-async function scrapeEspn(req, res) {
+async function scrapeEspn(startDate, endDate, req, res) {
```
Comment (saravpereira) [inline controllers/webscrapeController.js]: you are not using req in this function, needs to be removed from here and function call

File: controllers/webscrapeController.js [hunk 14]
```diff
@@ -1,21 +1,106 @@
...
+          const matchDateObj = new Date(
+            parseInt(dateString.substring(0, 4)),
+            parseInt(dateString.substring(4, 6)) - 1,
+            parseInt(dateString.substring(6, 8))
+          );
+          const options = { year: 'numeric', month: 'long', day: 'numeric' };
+          matchInfo.matchDate = matchDateObj.toLocaleDateString('en-US', options);
+          
+          const leagueElement = container.closest(".Card")?.querySelector(".Card__Header__Title");
+          matchInfo.league = leagueElement ? leagueElement.textContent.trim() : '';
```
Comment (saravpereira) [inline controllers/webscrapeController.js]: this can be rewritten as
 matchInfo.league = leagueElement?.textContent?.trim() ?? '';

File: models/dates.js [hunk 15]
```diff
@@ -0,0 +1,47 @@
+function getDateRange(userEndDate) {
```
Comment (saravpereira) [inline models/dates.js]: these are util functions.
models are files where we describe the schema type. ex: expected payload attributes for POST request.
create a separate folder called utils and move this dateUtils.js there

File: controllers/webscrapeController.js [hunk 16]
```diff
@@ -53,7 +53,7 @@ async function scrapeEspn(startDate, endDate, req, res) {
           matchInfo.matchDate = matchDateObj.toLocaleDateString('en-US', options);
           
           const leagueElement = container.closest(".Card")?.querySelector(".Card__Header__Title");
-          matchInfo.league = leagueElement ? leagueElement.textContent.trim() : '';
+          matchInfo.league = leagueElement?.textContent?.trim() ?? '';
```
Comment (saravpereira) [inline controllers/webscrapeController.js:56]: same changes should be applied below for consistency

File: controllers/webscrapeController.js [hunk 17]
```diff
@@ -1,21 +1,106 @@
...
+    
+          const awayLogoElement = container.querySelector(".ScoreboardScoreCell__Item--away .ScoreboardScoreCell__Logo");
+          matchInfo.awayLogo = awayLogoElement ? awayLogoElement.getAttribute("src") : ' ';
+
+          if (!matchInfo.homeLogo) {
+            matchInfo.homeLogo = 'https://www.seekpng.com/png/full/28-289657_espn-soccer-team-logo-default.png';
+          }
+        
+          if (!matchInfo.awayLogo) {
+            matchInfo.awayLogo = 'https://www.seekpng.com/png/full/28-28957_espn-soccer-team-logo-default.png';
```
Comment (saravpereira) [inline controllers/webscrapeController.js]: this is a bad url fyi. (put it in your browser and check, I think you miss a character)

also, all links need to be in the constants file for reusability ease

Comment (nghiapham1026) [reply controllers/webscrapeController.js]: > this is a bad url fyi. (put it in your browser and check, I think you miss a character)
> 
> also, all links need to be in the constants file for reusability ease

https://www.seekpng.com/png/full/28-289657_espn-soccer-team-logo-default.png
28-289657_espn-soccer-team-logo-default

**Comment by nghiapham1026** [comment]
https://www.seekpng.com/png/full/28-289657_espn-soccer-team-logo-default.png
The link works fine
28-289657_espn-soccer-team-logo-default

**Review by saravpereira — APPROVED** [review]
